### PR TITLE
Add test file

### PR DESCRIPTION
**Pull request by janEbert**
Correspondingly, add the generated "tags" file to .gitignore.
I tried to add "actual code" and have too many tests rather than only
complex ones.
Surely, some cases are not covered. For example, using multiple macros
for a definition.

### CODE REVIEW

**Comment by oxinabox** [comment]
Do you think we can construct some actual tests that run on CI?
They would need to be written in Bash, I guess.
Or in Julia using ```run(`ctag ...`)```
which could be fine

**Comment by janEbert** [comment]
Sure, I'd love to look into that! Though I have never set up CI.
The tests would fail though as with the simple regexes, edge cases are simply not covered:
```
#=
const old_zero = 0
=#
```
... would be parsed as an identifier.

That is also why I think a proper parser would be more useful. I have thought about writing one for Universal Ctags that would be more sophisticated than the current one but it would take some time. Guess the tests failing would be the exact thing we want until then. :)

**Comment by oxinabox** [comment]
I think easiest would be to setup CI for julia and write the tests in julia,
after doing `readlines` on the genated file.

Then you could do `@test_broken` for the edge cases that fail.

You can copy the .travisci from any julia project to do that.

**Comment by oxinabox** [comment]
If you write the tests, so you can run them locally
I will make it work with Travis CI

**Comment by janEbert** [comment]
How would you go about using the external [Universal Ctags](https://github.com/universal-ctags/ctags)? They don't have release tags and the binary name may be hard to find due to name overlaps with Exuberant Ctags or Etags.

Or should I just write, try and see using `ctags` as the executable?

> If you write the tests, I will make it work with Travis CI

Thanks, but I'd actually be glad to learn by doing!

**Comment by oxinabox** [comment]
Can't we just use Exuberant Ctags ?
Or are we using new features? (if so that needs to go in the readme)

**Comment by janEbert** [comment]
You are absolutely right and I see why naming is no problem.
We do not use new features, everything is still vanilla. :)

I added the tests and Travis CI, hope you are content with the code.
I tried to make local testing easy as well by trying different binaries.

Broken features are marked depending on the identifier name or the "kind" (function, variable, struct, ...).

The tests will fail now because I based them on the updated ctags file (#4).

File: Project.toml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+name = "julia-ctags"
+uuid = "106e64ca-b22d-11e9-257b-81adddf2616c"
+authors = ["janEbert <janpublicebert@posteo.net>"]
+version = "0.1.0"
```
Comment (oxinabox) [inline Project.toml]: I think we can delete moist of this except the `name` field,
since we won't be registering this

**Comment by janEbert** [comment]
Hey, the build keeps being stuck on "Resolving package versions..." when testing. I couldn't find any other reports via searching and have checked out other `Project.toml`s and `runtest.jl`s, however, nothing seems different.
Do you have an idea what could be wrong?

File: test/runtests.jl [hunk 2]
```diff
@@ -0,0 +1,14 @@
+using Test
+
+include(joinpath(@__DIR__, "test_tags.jl"))
```
Comment (oxinabox) [inline test/runtests.jl]:
```suggestion
include( "test_tags.jl")
```

Include (when used at toplevel) is alway a relative path

**Comment by oxinabox** [comment]
Do your tests work fine locally?

**Comment by janEbert** [comment]
Yes, I tested just now with:
```
$ JULIA_LOAD_PATH=$PWD julia
julia> ]
(julia-ctags) pkg> test
```
And it worked fine. Though as I mentioned, _if_ the tests ran in Travis, they would fail because I based them on the new `ctags` file in #4.

Edit:
I also just tested using the same command Travis uses (plus the `--project` flag) and it did not hang up.

**Comment by janEbert** [comment]
Sorry to spam; after trying a lot, I still haven't figured out how to remove those duplicate commits from `master`. I will fix it later, though.

**Comment by karlwessel** [comment]
> Hey, the build keeps being stuck on "Resolving package versions..." when testing. I couldn't find any other reports via searching and have checked out other `Project.toml`s and `runtest.jl`s, however, nothing seems different.
> Do you have an idea what could be wrong?

I tried your PR at my own machine and it gets stuck in an endless loop because none of the predefined ctags executables worked. The same probably happens to Travis. 

You probably should make sure that `trygeneratetags` and the tests exit gracefully if no ctags executable is found and maybe add `ctags` to the list of ctags executables in `ctagsbins`.

File: test/test_tags.jl [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            try
+                tags = generatetags(testfile, ctagsbins[ctagsbinindex])
+                success = true
+            catch LoadError
+                ctagsbinindex += 1
+            end
+        end
+    end
+    return tags
+end
```
Comment (oxinabox) [inline test/test_tags.jl:63]: I think this should be:

```
function trygeneratetags(testfile::AbstractString)
    for ctag_binary in ctagsbins
        try
            return generatetags(testfile, ctag_binary)
        catch err
            # for compat with Julia version accept either exception type
            err isa ErrorException || err isa ProcessFailedException  || rethrow()
       end
end
```

**Comment by m-wells** [comment]
I'm not sure how much packaging you want to do with this but you could always provide a ctags build if one wasn't found either with [BinaryProvider.jl](https://github.com/JuliaPackaging/BinaryProvider.jl) or with the [Anaconda Ctags package](https://anaconda.org/conda-forge/ctags) using [Conda.jl](https://github.com/JuliaPy/Conda.jl).

If you go this route you would probably want to make this into a proper julia package.

**Comment by oxinabox** [comment]
>  If you go this route you would probably want to make this into a proper julia package.

That wouldn't really make sense.
This file needs to go into your ctags config

**Comment by m-wells** [comment]
> > If you go this route you would probably want to make this into a proper julia package.
> 
> That wouldn't really make sense.
> This file needs to go into your ctags config

Not really.  It can be passed in as an option to the `ctags` command via `--options=pathname`

**Comment by oxinabox** [comment]
When you do `Pkg.add` or even use Artifacts stuff the files endup in an arbitrary location.
Not somewhere you can reliably find them.

**Comment by m-wells** [comment]
@oxinabox I created a pull request #6 to package julia-ctags into `JuliaCtags` (to stick with the Julia package naming conventions).  It isn't much but it provides a method to copy the `ctags` config file into the current directory (naming the new file `julia_ctags` by default).  If a `ctags` install was shipped with this package (see my comment in the pull request) other methods could be implemented that allow users to generate tags files using the package.

In addition, optionally providing a `ctags` install also means testing this package becomes much easier on Travis (or any other CI platform).

**Comment by janEbert** [comment]
Hey Karl, thank you so much for finding that. I actually hadn't forgotten this PR but due to the CI failing and me being so ashamed of the test code (it doesn't seem "right" at all), I never cared finishing it... Sorry! :sweat_smile:

I will fix it later so #6 can be merged as well. By the way, I also wrote an Etags file for use with the Emacs-internal tags engine. Do you want it in the repo as well?

**Comment by janEbert** [comment]
Sorry it took so long. I also remembered why I didn't add the `ctags` executable in the first place. It was to avoid incompatibilities in case `etags`, often mapped to `ctags`, was found. Of course, due to the possibility of simply ordering the binaries, that didn't make sense. Also, that `while`-loop was pretty embarrassing, oh man... Good it's gone now. :)

File: test/test_tags.jl [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+"Path to Ctags configuration file (the parser)."
+const ctagsconfigpath = joinpath(@__DIR__, "..", "ctags")
+"Test file to generate the tags from."
+const testfilepath = joinpath(@__DIR__, "testfile.jl")
+"File containing test tags to compare against."
+const testtagspath = joinpath(@__DIR__, "testtags")
+
+"Ctags binaries to sequentially try to call if one is not found."
+const ctagsbins = (
+        "uctags",
```
Comment (karlwessel) [inline test/test_tags.jl:21]: You should change the order of the possible tag command in such a way that the tag command available at your own system is last in the list. This way your make sure that you test the error case of the `trygeneratetags` method.

It would probably also be a good idea to only add those tag binaries to the list for which you can actually execute the tests. It is better to have a short but testes list of binaries then a complete but untested list.

People using binaries not in the list can then add their binary to the list them self, see if the test work and create an according PR or issue if necessary.

File: test/test_tags.jl [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    return converttags(tags)
+end
+
+function trygeneratetags(testfile::AbstractString)
+    for ctagsbinary in ctagsbins
+        try
+            return generatetags(testfile, ctagsbinary)
+        catch e
+            # For compatibility accept either exception type.
+            if !(err isa LoadError || err isa ErrorException
```
Comment (karlwessel) [inline test/test_tags.jl:57]: For uctags the `generatetags` method failed with a `Base.IOError` instead of a `LoadError` (although from the printed error message it looked like a `LoadError`), so you should add that one to the list of 'acceptable' error.

I'm not sure if adding `LoadError` to the list of acceptable Exceptions is necessary. At least for my setup none of the failing tag binaries did throw an actual `LoadError`.

**Review by karlwessel — CHANGES_REQUESTED** [review]
Thanks for taking the time to write those tests! I added a few comments on the current commit.

File: test/test_tags.jl (hunk 5 above)
Comment (oxinabox) [reply test/test_tags.jl:57]: I expected `ProcessFailedException`

Comment (janEbert) [reply test/test_tags.jl:57]: Sorry, I was busy when pushing and didn't test. I'll be sure to fix it, thanks again, Karl!
I actually can't remember why I included the `LoadError` (should have documented that one) but I'm sure it's there for a reason. Maybe I can replicate it again. But yeah if the `IOError` looked like a `LoadError` that may be the reason.

Hey Lyndon, that one's listed one line below, you can't see it in the diff.

Comment (oxinabox) [reply test/test_tags.jl:57]: ah cool cool
